README.md
---
title: Screen Orientation
description: Set the screen orientation
---
<!--
# license: Licensed to the Apache Software Foundation (ASF) under one
#         or more contributor license agreements.  See the NOTICE file
#         distributed with this work for additional information
#         regarding copyright ownership.  The ASF licenses this file
#         to you under the Apache License, Version 2.0 (the
#         "License"); you may not use this file except in compliance
#         with the License.  You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#         Unless required by applicable law or agreed to in writing,
#         software distributed under the License is distributed on an
#         "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#         KIND, either express or implied.  See the License for the
#         specific language governing permissions and limitations
#         under the License.
-->

# Cordova Screen Orientation Plugin

[![Android Testsuite](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/android.yml/badge.svg)](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/android.yml) [![Chrome Testsuite](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/chrome.yml/badge.svg)](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/chrome.yml) [![iOS Testsuite](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/ios.yml/badge.svg)](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/ios.yml) [![Lint Test](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/lint.yml/badge.svg)](https://github.com/apache/cordova-plugin-screen-orientation/actions/workflows/lint.yml)

Cordova plugin to set/lock the screen orientation in a common way for iOS, Android, and windows-uwp.  This plugin is based on [Screen Orientation API](http://www.w3.org/TR/screen-orientation/) so the api matches the current spec.

The plugin adds the following to the screen object (`window.screen`):

```js
// lock the device orientation
.orientation.lock('portrait')

// unlock the orientation
.orientation.unlock()

// current orientation
.orientation
```

## Install


```bash
cordova plugin add cordova-plugin-screen-orientation
```

## Supported Orientations

#### portrait-primary
> The orientation is in the primary portrait mode.

#### portrait-secondary
> The orientation is in the secondary portrait mode.

#### landscape-primary
> The orientation is in the primary landscape mode.

#### landscape-secondary
> The orientation is in the secondary landscape mode.

#### portrait
> The orientation is either portrait-primary or portrait-secondary (sensor).

#### landscape
> The orientation is either landscape-primary or landscape-secondary (sensor).

#### any
>  orientation is  unlocked - all orientations are supported.

## Usage

```js
// set to either landscape
screen.orientation.lock('landscape');

// allow user rotate
screen.orientation.unlock();

// access current orientation
console.log('Orientation is ' + screen.orientation.type);
```

## Events

Both android and iOS will fire the orientationchange event on the window object.
For this version of the plugin use the window object if you require notification.


### Example usage

```js
window.addEventListener("orientationchange", function(){
    console.log(screen.orientation.type); // e.g. portrait
});
```

The 'change' event listener has also been added to the screen.orientation object.

### Example usage

```js
screen.orientation.addEventListener('change', function(){
    console.log(screen.orientation.type); // e.g. portrait
});
    // OR

screen.orientation.onchange = function(){console.log(screen.orientation.type);
};

```
## Android Notes

The __screen.orientation__ property will not update when the phone is [rotated 180 degrees](http://www.quirksmode.org/dom/events/orientationchange.html).

## Windows UWP Notes

Windows store apps (windows-uwp) will only display orientation changes if the device has some sort of accelerometer.  The internal state of the "orientation" will still be kept, but the actual screen won't rotate unless the device supports it.


RELEASENOTES.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->

# Release Notes

### 3.0.4 (Oct 27, 2023)

**Fixes:**

* [GH-116](https://github.com/apache/cordova-plugin-screen-orientation/pull/116) fix(android, ios): use clobbers to overwrite `screen.orientation`

**Others:**

* [GH-125](https://github.com/apache/cordova-plugin-screen-orientation/pull/125) dep: bump `word-wrap` from `1.2.3` to `1.2.5`
* [GH-124](https://github.com/apache/cordova-plugin-screen-orientation/pull/124) chore: update asf config
* [GH-123](https://github.com/apache/cordova-plugin-screen-orientation/pull/123) ci(gh-action): sync with paramedic configs

### 3.0.3 (Feb 21, 2023)
* build(deps): bump json5 from 1.0.1 to 1.0.2 ([#110](https://github.com/apache/cordova-plugin-screen-orientation/issues/110))
* feat(ios): add support for **iOS** 16 ([#107](https://github.com/apache/cordova-plugin-screen-orientation/issues/107))
* fix(chrome): lock test ([#105](https://github.com/apache/cordova-plugin-screen-orientation/issues/105))
* chore: add .npmrc file to ensure the official NPM registry is used for package installations. ([#103](https://github.com/apache/cordova-plugin-screen-orientation/issues/103))
* chore: package-lock update (#104)
* ci: sync workflow with paramedic (#101)
* chore(npm): bump package-lock v2 w/ rebuild (#99)
* ci(android): update java requirement for `cordova-android`@11 ([#92](https://github.com/apache/cordova-plugin-screen-orientation/issues/92))
* ci(ios): update workflow w/ **iOS** 15 ([#88](https://github.com/apache/cordova-plugin-screen-orientation/issues/88))
* ci: add action-badge (#87)
* ci: remove travis & appveyor (#86)
* ci: add gh-actions workflows ([#84](https://github.com/apache/cordova-plugin-screen-orientation/issues/84))
* ci: add node-14.x to workflow ([#76](https://github.com/apache/cordova-plugin-screen-orientation/issues/76))
* fix: resolveOrientation function not working correctly ([#77](https://github.com/apache/cordova-plugin-screen-orientation/issues/77))
* ci(travis): update osx xcode image (#72)
* ci(travis): updates **Android** API level (#71)
* chore: adds package-lock file ([#67](https://github.com/apache/cordova-plugin-screen-orientation/issues/67))
* refactor(eslint): use cordova-eslint /w fix ([#66](https://github.com/apache/cordova-plugin-screen-orientation/issues/66))
* chore(npm): use short notation in `package.json` ([#64](https://github.com/apache/cordova-plugin-screen-orientation/issues/64))
* chore: removes old demo project. closes #52 (#65)
* chore(asf): update git notification settings
* Update CONTRIBUTING.md
* chore(npm): improve ignore list ([#60](https://github.com/apache/cordova-plugin-screen-orientation/issues/60))
* ci: updates Node.js versions ([#61](https://github.com/apache/cordova-plugin-screen-orientation/issues/61))
* test: let Jasmine handle promises ([#57](https://github.com/apache/cordova-plugin-screen-orientation/issues/57))
* ci(appveyor): replace node 6 with node 12
* ci(travis): upgrade to node 8
* chore(release): 3.0.3-dev

### 3.0.2 (Jun 19, 2019)

-   fix(ios): Landscape Issue Fix and prevent orientation change when already in orientation ([#25](https://github.com/apache/cordova-plugin-screen-orientation/issues/25)) ([`a89e94c`](https://github.com/apache/cordova-plugin-screen-orientation/commit/a89e94c), [`13b34f0`](https://github.com/apache/cordova-plugin-screen-orientation/commit/13b34f0))
-   build: update `.npmignore` to remove unneeded files from npm package ([`80724df`](https://github.com/apache/cordova-plugin-screen-orientation/commit/80724df))
-   build: add `.gitattributes` to force LF (instead of possible CRLF on Windows) ([`ae60fa9`](https://github.com/apache/cordova-plugin-screen-orientation/commit/ae60fa9))
-   docs: move legacy release notes to `RELEASENOTES.md` ([`34c9c29`](https://github.com/apache/cordova-plugin-screen-orientation/commit/34c9c29))
-   ci(travis): Update Travis CI configuration for new paramedic ([#47](https://github.com/apache/cordova-plugin-screen-orientation/issues/47)) ([`83370a8`](https://github.com/apache/cordova-plugin-screen-orientation/commit/83370a8))
-   chore: add missing keywords to package.json for plugin search ([#49](https://github.com/apache/cordova-plugin-screen-orientation/issues/49)) ([`7f493ce`](https://github.com/apache/cordova-plugin-screen-orientation/commit/7f493ce))
-   ci: drop Node.js v4 support ([#45](https://github.com/apache/cordova-plugin-screen-orientation/issues/45)) ([`de7f6fe`](https://github.com/apache/cordova-plugin-screen-orientation/commit/de7f6fe))
-   chore(github): Add or update GitHub pull request and issue template ([`ccc17fb`](https://github.com/apache/cordova-plugin-screen-orientation/commit/ccc17fb))
-   ci(travis): also accept terms for android sdk `android-27` ([`9d2a422`](https://github.com/apache/cordova-plugin-screen-orientation/commit/9d2a422))
-   chore: [CB-11843](https://issues.apache.org/jira/browse/CB-11843) fix package information ([#33](https://github.com/apache/cordova-plugin-screen-orientation/issues/33)) ([`950d339`](https://github.com/apache/cordova-plugin-screen-orientation/commit/950d339))

www/screenorientation.js
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

/* global cordova, OrientationLockType */

const screenOrientation = {};
if (!window.OrientationType) {
    window.OrientationType = {
        'portrait-primary': 0,
        'portrait-secondary': 180,
        'landscape-primary': 90,
        'landscape-secondary': -90
    };
}
if (!window.OrientationLockType) {
    window.OrientationLockType = {
        'portrait-primary': 1,
        'portrait-secondary': 2,
        'landscape-primary': 4,
        'landscape-secondary': 8,
        portrait: 3, // either portrait-primary or portrait-secondary.
        landscape: 12, // either landscape-primary or landscape-secondary.
        any: 15 // All orientations are supported (unlocked orientation)
    };
}
let orientationMask = 1;
screenOrientation.setOrientation = function (orientation) {
    orientationMask = window.OrientationLockType[orientation];
    cordova.exec(null, null, 'CDVOrientation', 'screenOrientation', [orientationMask, orientation]);
};

screenOrientation.lock = function (orientation) {
    const p = new Promise(function (resolve, reject) {
        resolveOrientation(orientation, resolve, reject);
    });
    return p;
};

screenOrientation.unlock = function () {
    screenOrientation.setOrientation('any');
};

setOrientationProperties();

function resolveOrientation (orientation, resolve, reject) {
    if (!Object.prototype.hasOwnProperty.call(OrientationLockType, orientation)) {
        const err = new Error();
        err.name = 'NotSupportedError';
        reject(err); // "cannot change orientation");
    } else {
        screenOrientation.setOrientation(orientation);
        resolve('Orientation set'); // orientation change successful
    }
}

let onChangeListener = null;

Object.defineProperty(screenOrientation, 'onchange', {
    set: function (listener) {
        if (onChangeListener) {
            screenOrientation.removeEventListener('change', onChangeListener);
        }
        onChangeListener = listener;
        if (onChangeListener) {
            screenOrientation.addEventListener('change', onChangeListener);
        }
    },
    get: function () {
        return onChangeListener || null;
    },
    enumerable: true
});

const evtTarget = new XMLHttpRequest(); // document.createElement('div');
const orientationchange = function () {
    setOrientationProperties();
    const event = document.createEvent('Events');
    event.initEvent('change', false, false);
    evtTarget.dispatchEvent(event);
};

screenOrientation.addEventListener = function (a, b, c) {
    return evtTarget.addEventListener(a, b, c);
};

screenOrientation.removeEventListener = function (a, b, c) {
    return evtTarget.removeEventListener(a, b, c);
};

function setOrientationProperties () {
    switch (window.orientation) {
    case 0:
        screenOrientation.type = 'portrait-primary';
        break;
    case 90:
        screenOrientation.type = 'landscape-primary';
        break;
    case 180:
        screenOrientation.type = 'portrait-secondary';
        break;
    case -90:
        screenOrientation.type = 'landscape-secondary';
        break;
    default:
        screenOrientation.type = 'portrait-primary';
        break;
    }
    screenOrientation.angle = window.orientation || 0;
}
window.addEventListener('orientationchange', orientationchange, true);

module.exports = screenOrientation;
